### PR TITLE
[5.1] Example of recently inserted state tracking

### LINKED ISSUE #9745: During insert, set a flag on the model
When doing a `firstOrCreate`, it comes up frequently that I'd like to return a 201 when the action resulted in an insert.  It would be nice if a flag existed on model subclasses that allowed me to infer if a model with  `exists == true` was *also* created during the current request.

### PR DESCRIPTION
**Pull request by atrauzzi**
Fairly trivial, but I think this could be helpful in many scenarios.

Closes #9745.

### CODE REVIEW

**Comment by Anahkiasen** [comment]
That would be very useful

**Comment by atrauzzi** [comment]
And of course, anyone is open to suggest a better name for the variable ;)

File: src/Illuminate/Database/Eloquent/Model.php [hunk 1]
```diff
@@ -187,6 +187,13 @@
     public $exists = false;
 
     /**
+     * Indicates if the model was inserted during the current request/run.
+     * 
+     * @var bool
+     */
+     public $recent = false;
```
Comment (brunogaspar) [inline src/Illuminate/Database/Eloquent/Model.php]: Minor indentation issue here

**Comment by brunogaspar** [comment]
I like the idea.

**Comment by taylorotwell** [comment]
What were the use cases for this?

**Comment by deefour** [comment]
If I understand correctly, this would help for places we don't want to hook some `created` event in an observer; where inlining the conditional functionality is just fine.

```php
function persist(Podcast $podcast, array $attrs) {
  $podcast = $podcast ?: new Podcast;
  
  $podcast->forceFill(array_only($attrs, [ 'title', 'url' ]));
  
  $podcast->save();
  
  if ($podcast->recent) {
    // do something with $podcast that required it
    // already being saved to the db, but only when
    // it was just created.
  }
}
```

Setting a flag up front is currently necessary.

```php
function persist(Podcast $podcast, array $attrs) {
  $podcast  = $podcast ?: new Podcast;
  $creating = ! $podcast->exists; // Need to set a flag here; this PR makes this unecessary
  
  $podcast->forceFill(array_only($attrs, [ 'title', 'url' ]));
  
  $podcast->save();
  
  if ($creating) {
    // ...
  }
}
```

**Comment by atrauzzi** [comment]
Without functionality along these lines, repetitive blocks using `firstOrNew` happen.  In our case, we need it to respond with HTTP 200 or 201.

With this, we would be looking at `return new JsonResponse($model, $model->recent ? 201 : 200);`

**Comment by lukasgeiter** [comment]
I think the variable could be a bit more self-explanatory. What about `recentlyCreated`?

**Comment by atrauzzi** [comment]
@lukasgeiter - I'd like to see if there is a different name that is more specific, although it might be nice to keep it as short as possible.

**Comment by GrahamCampbell** [comment]
This feature would be really useful to me too. :+1:

**Comment by GrahamCampbell** [comment]
My use case is working out if a user has signed up.

**Comment by dtaub** [comment]
What about `new` for the variable name? It's concise and (imo) better conveys the fact that the record was newly created. `recent` almost sounds like it's checking for some arbitrary age, whereas `new` implies it was created just now on the current request.

**Comment by Anahkiasen** [comment]
Isn't `new` a reserved keyword though?

**Comment by GrahamCampbell** [comment]
> Isn't new a reserved keyword though?

No, not for properties. I tried it out too just to be sure.

**Comment by Anahkiasen** [comment]
Then in that case I'd prefer `new` too I think, better conveys the intent than `recent`

**Comment by lukasgeiter** [comment]
:+1: for `new`. But what would its initial value be when you create a fresh model instance?

**Comment by GrahamCampbell** [comment]
Yeh.

**Comment by dtaub** [comment]
@Anahkiasen Yeah, I was thinking it might be reserved, too, but double checked before posting. The only issue that crossed my mind was that to someone used to Ruby, it might kinda look like it's creating a new instance, but that's kind of a stretch.

**Comment by arrilot** [comment]
I would have gone even further - `isNew` because we already have a verb `exists` 

```php
if ($model->exists) {
...
}

if ($model->isNew) {
...
}
```

**Comment by dan-har** [comment]
I think it is bad to add more public properties to the model, it can make conflicts with the dynamic properties, for example if i have a column 'new' or 'recent' inside a table. It should be protected a property that is accessed by a public method.

**Comment by dtaub** [comment]
Yeah, it would be pretty standard to have a protected property `new` and access it using `isNew()`. No point in making it public.

**Comment by atrauzzi** [comment]
@GrahamCampbell @taylorotwell -- Let me know what you guys would prefer to see and I'll amend my MR.

**Comment by GrahamCampbell** [comment]
@atrauzzi It's totally up to @taylorotwell. :)

**Comment by atrauzzi** [comment]
For sure, @taylorotwell - Whatcha think? :)

**Comment by taylorotwell** [comment]
Renamed it to `wasRecentlyCreated` to make it a little longer and less likely to collide.

**Comment by atrauzzi** [comment]
:+1: Awesome!

**Comment by hauthorn** [comment]
I'm not sure to write this here.
This addition does make some testing using Laravels "assertViewHas" fail tests.

See my gist here for an explanation: https://gist.github.com/hauthorn/fb1494c3ba5238283acc

This didn't fail before. Now it does. 

Do I use the "assertViewHas" in some inappropriate way, or is this an unintended consequence of the change?

**Pull request merged**